modules/ximgproc/src/paillou_filter.cpp

```diff
@@ -0,0 +1,449 @@
+        {
+            for (int j = range.start; j<range.end; j++)
+            {
+                // Equation 26 p194
+                short *c1 = (short*)img.ptr(0) + j;
+                f2 = ((float*)im1.ptr(0)) + j;
+                double border = *c1;
+                yp[0] = *c1;
+                c1 += cols;
+                yp[1] = *c1 - b1*yp[0] - b2*border;
+                c1 += cols;
+                for (int i = 2; i<rows; i++, c1 += cols)
+                    yp[i] = *c1 - b1*yp[i - 1] - b2*yp[i - 2];
+                // Equation 27 p194
+                c1 = (short*)img.ptr(rows - 1) + j;
+                border = *c1;
+                ym[rows - 1] = *c1;
+                c1 -= cols;
+                ym[rows - 2] = *c1 - b1*ym[rows - 1];
+                c1 -= cols;
+                for (int i = rows - 3; i >= 0; i--, c1 -= cols)
+                    ym[i] = *c1 - b1*ym[i + 1] - b2*ym[i + 2];
+                // Equation 25 p193
+                for (int i = 0; i<rows; i++, f2 += cols)
+                    *f2 = (float)(a1*(ym[i] - yp[i]));
+            }
+        }
+        break;
+        case CV_16U :
+        {
+            for (int j = range.start; j<range.end; j++)
+            {
+                // Equation 26 p194
+                unsigned short *c1 = (unsigned short*)img.ptr(0) + j;// c1 pointeur sur x(i-1)
+                f2 = ((float*)im1.ptr(0)) + j;
+                double border = *c1;
+                yp[0] = *c1;
+                c1 += cols;
+                yp[1] = *c1 - b1*yp[0] - b2*border;
+                c1 += cols;
+                for (int i = 2; i<rows; i++, c1 += cols)
+                    yp[i] = *c1 - b1*yp[i - 1] - b2*yp[i - 2];
+                // Equation 27 p194
+                c1 = (unsigned short*)img.ptr(rows - 1) + j;
+                border = *c1;
+                ym[rows - 1] = *c1;
+                c1 -= cols;
+                ym[rows - 2] = *c1 - b1*ym[rows - 1];
+                c1 -= cols;
+                for (int i = rows - 3; i >= 0; i--, c1 -= cols)
+                    ym[i] = *c1 - b1*ym[i + 1] - b2*ym[i + 2];
+                // Equation 25 p193
+                for (int i = 0; i<rows; i++, f2 += cols)
+                    *f2 = (float)(a1*(ym[i] - yp[i]));
+            }
+        }
+            break;
+        default :
+            return ;
+            }
+    };
+    ParallelGradientPaillouYCols& operator=(const ParallelGradientPaillouYCols &) {
+         return *this;
+    };
+};
+
+
+class ParallelGradientPaillouYRows: public ParallelLoopBody
+{
+private:
+    Mat &img;
+    Mat &dst;
+    double a;
+    double w;
+    bool verbose;
+
+public:
+    ParallelGradientPaillouYRows(Mat& imgSrc, Mat &d,double aa,double ww):
+        img(imgSrc),
+        dst(d),
+        a(aa),
+        w(ww),
+        verbose(false)
+    {}
+    void Verbose(bool b){verbose=b;}
+    virtual void operator()(const Range& range) const
+    {
+        CV_Assert(img.depth()==CV_32FC1);
```

Comment:
You allow more Mat types inside ParallelGradientPaillouYCols than here. Is it correct?